examples/solidity/ERC20/erc20.qnt

```diff
@@ -1,246 +1,359 @@
-    val example = all {
-      lastTx.tag == "approve",
-      not(lastTx.failed),
-      lastTx.value > 0,
-    }
-
-    not(example)
-
-  /* EXPECTED PROPERTIES */
-
-  /* No transferFrom should be possible, while there is a pending approval
-    for a smaller amount. This invariant is violated, as explained in:
-    https://docs.google.com/document/d/1YLPtQxZu1UAvO9cZ1O2RPXBbT0mooh4DYKjA_jp-RLM/edit# */
-
-  val NoTransferFromWhileApproveInFlight =
-    val badExample = all {
-      lastTx.tag == "transferFrom",
-      lastTx.value > 0,
-      not(lastTx.failed),
-      nondet tx = oneOf(pendingTransactions)
-      all {
-        tx.tag == "approve",
-        tx.sender == lastTx.fromAddr,
-        tx.spender == lastTx.sender,
-        tx.value < lastTx.value,
-        tx.value > 0,
-      }
-    }
-    not(badExample)
-
-  /* Make sure that the account balances never go negative. */
-  val NoNegativeBalances = ADDR.forall(a => balanceOf.get(a) >= 0)
+    }
+
+    /**
+     * ERC20: Moves amount tokens from the caller’s account to to.
+     * Returns a boolean value indicating whether the operation succeeded.
+     *
+     * Quint: also return an error code and the new state.
+     * If the error != "", the new state should be applied.
+     */
+    pure def transfer(state: Erc20State, sender: Address,
+            toAddr: Address, amount: Uint): (bool, Erc20Result) = {
+        // `transfer` always returns true, but we should check Erc20Result.error
+        (true, _transfer(state, sender, toAddr, amount))
+    }
+
+    /**
+     * ERC20: Returns the remaining number of tokens that spender will be allowed to
+     * spend on behalf of owner through transferFrom. This is zero by default.
+     *
+     * This value changes when approve or transferFrom are called.
+     */
+    pure def allowance(state: Erc20State, owner: Address, spender: Address): Uint = {
+        state.allowance.get((owner, spender))
+    }
+
+    /**
+     * ERC20: Sets amount as the allowance of spender over the caller’s tokens.
+     *
+     * Returns a boolean value (and the new state) indicating whether the
+     * operation succeeded.
+     *
+     * Quint: also return an error code and the new state.
+     * If the error != "", the new state should be applied.
+     */
+    pure def approve(state: Erc20State, sender: Address,
+            spender: Address, amount: Uint): (bool, Erc20Result) = {
+         val err = require(sender != "", "ERC20: transfer from the zero address")
+            .andRequire(spender != "", "ERC20: transfer to the zero address")
+        if (err != "") {
+            (true, returnError(err, state))
+        } else {
+            // the case of sender == spender seems to be allowed
+            val newAllowance = state.allowance.set((sender, spender), amount)
+            (true, returnState(state.with("allowance", newAllowance)))
+        }
+    }
+
+    /**
+     * Moves amount tokens from from to to using the allowance mechanism.
```

Comment:
idk what "from from to to" means.

Did you mean from `fromAddr` to `toAddr`?

Reply:
Yeah, that's the ERC20 docs. Will fix it :)